File: src/main/java/rlbot/manager/BotLoopRenderer.java
package rlbot.manager;

import com.google.flatbuffers.FlatBufferBuilder;
import rlbot.Bot;
import rlbot.cppinterop.RLBotDll;
import rlbot.render.RenderPacket;
import rlbot.render.Renderer;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * A renderer that will be managed automatically. Users of this renderer should expect the packet
 * to be started and finished on a regular basis without their intervention.
 */
public class BotLoopRenderer extends Renderer {

    private static Map<Integer, BotLoopRenderer> botLoopMap = new ConcurrentHashMap<>();
    private RenderPacket previousPacket = null;

    private BotLoopRenderer(int index) {
        super(index);
    }

    public static BotLoopRenderer forBotLoop(final Bot bot) {
        botLoopMap.computeIfAbsent(bot.getIndex(), BotLoopRenderer::new);
        return botLoopMap.get(bot.getIndex());
    }

    void startPacket() {
        builder = new FlatBufferBuilder(1000);
    }

    void finishAndSendIfDifferent() {
        RenderPacket packet = doFinishPacket();
        if (!packet.equals(previousPacket)) {
            RLBotDll.sendRenderPacket(packet);
            previousPacket = packet;
        }
    }
}

File: src/main/cs/RLBotDotNet/RLBotDotNet/Renderer/BotLoopRenderer.cs
using FlatBuffers;

namespace RLBotDotNet.Renderer
{
    /// <summary>
    /// Renderer for the bot loop.
    /// </summary>
    public class BotLoopRenderer : Renderer
    {
        private RenderPacket _previousPacket;

        internal BotLoopRenderer(int index) : base(index)
        {
        }

        /// <summary>
        /// Starts the render packet.
        /// </summary>
        public void StartPacket()
        {
            Builder = new FlatBufferBuilder(1000);
        }

        /// <summary>
        /// Finishes the render packet and sends it if it's different from the last packet.
        /// </summary>
        public void FinishAndSendIfDifferent()
        {
            var packet = DoFinishPacket();
            if (!packet.Equals(_previousPacket))
            {
                SendPacket(packet);
                _previousPacket = packet;
            }
        }
    }
}